esp-hal/src/timer/systimer.rs

```diff
@@ -317,50 +311,49 @@ pub trait Unit {
         }
     }
 
-    /// Update the value returned by [Self::poll_count] to be the current value
-    /// of the counter.
-    ///
-    /// This can be used to read the current value of the timer.
-    fn update(&self) {
-        let systimer = unsafe { &*SYSTIMER::ptr() };
-        systimer
-            .unit_op(self.channel() as _)
-            .modify(|_, w| w.update().set_bit());
-    }
+    /// Reads the current counter value.
+    fn read_count(&self) -> u64 {
+        // This can be a shared reference as long as this type isn't Sync.
 
-    /// Return the count value at the time of the last call to [Self::update].
-    ///
-    /// Returns None if the update isn't ready to read if update has never been
-    /// called.
-    fn poll_count(&self) -> Option<u64> {
-        let systimer = unsafe { &*SYSTIMER::ptr() };
-        if systimer
-            .unit_op(self.channel() as _)
-            .read()
-            .value_valid()
-            .bit_is_set()
-        {
-            let unit_value = systimer.unit_value(self.channel() as _);
+        // The flag exists to ensure that we only update if the unit is not being
+        // accessed. If we always updated, we could end up wrecking another
+        // core's reading. There are two modes of failure:
+        // - on single core, an update in an interrupt handler races with readin lo and
+        //   hi.
+        // - on multicore, the other core can start an update after we already stopped
+        //   looping for valid.
+
+        // Perf impact:
+        // - ESP32 has no systimer, no impact
+        // - ESP32-S3 and P4 have atomic instructions, no multicore critical section
+        //   should be needed.
+        // - Other chips are single core, the atomic swap chould acquire a single-core
+        //   critical section which should be cheap enough.
+        let flag = self.updating_flag();
+        let channel = self.channel() as usize;
+        let was_updating = flag.swap(true, Ordering::Acquire);
```

Comment:
I can't believe I'm saying this but a full blown critical section might be more desirable here. I'm concerned an InterruptExecutor setup could cause issues with this implementation.

Reply:
Oh actually, perhaps we can use one unit per core, that way we can get away with a plain interrupt::free rather than a critical section. 🤔

Reply:
Prove to me that this is incorrect and I'll happily put a critical section into this function. Right now I feel like I've put enough effort into making this correct, and a critical section is fairly expensive on multi-core just to ensure the invariants we need here.

Reply:
So let's say we have two tasks, task A (high priority) and task B (low priority).

Let's say task B:
- Sets the atomic
- asks for an update
- gets prempted by task A

Task A also wants the time: 
- checks the atomic
- skips the update
- polls until it gets a count
- does CPU intensive work for 5 seconds
- yields (waits for interrupt I guess) 

Task B now resumes: 
- polls and immediately gets a count (since task A has done the polling)
- task B now has a count that is 5 seconds stale
- task B unsets the atomic

Of course, if there are other tasks with higher priority that prempt task B before it unsets the atomic, they will all get stale (and unchanged) counts until task B resumes.

I also just realised this could result in a dead lock if the higher priority task is blocking until the count increases to a certain amount, i.e. `Delay::new().delay_ms(10)`.

Reply:
Yaay, good point, thanks. Also the tests fail which I hope is caused by this because I'm pretty baffled right now :)

Reply:
I'm thinking we might be able to get away with `interrupt::free` and a counter (to prevent the first core from looping back into updating inappropriately). Instead of the `swap(true)` we'd `fetch_and_add(1)`. The interrupt free context ensures the counter is only ever 0, 1 or 2. The problem I can see is, two cores reading the counter in a loop can mutually prevent each other from updating, which is a weird edge case, one which I'm not sure we have to handle correctly.

Reply:
> task B now has a count that is 5 seconds stale

To be quite fair, nothing really prevents this if that blocking interrupt handler is fired. The critical section is released, the interrupt is handled, and the lower priority context still ends up with a 5 seconds stale counter. The deadlock is a much better argument.

Reply:
> Anyway I'm not sure we'd actually end up with something faster.

Yeah using interrupt::free only makes sense performance wise if you skip the atomics on the counter/flag.

Reply:
Why only then? How else would you guarantee consistency on dualcores?